File: src/main/java/com/eager/questioncloud/domain/verification/repository/EmailVerificationRepositoryImpl.java
package com.eager.questioncloud.domain.verification.repository;

import static com.eager.questioncloud.domain.verification.entity.QEmailVerificationEntity.emailVerificationEntity;

import com.eager.questioncloud.domain.user.entity.UserEntity;
import com.eager.questioncloud.domain.user.model.User;
import com.eager.questioncloud.domain.verification.dto.EmailVerificationWithUser;
import com.eager.questioncloud.domain.verification.entity.EmailVerificationEntity;
import com.eager.questioncloud.domain.verification.model.EmailVerification;
import com.eager.questioncloud.domain.verification.vo.EmailVerificationType;
import com.eager.questioncloud.exception.CustomException;
import com.eager.questioncloud.exception.Error;
import com.querydsl.jpa.impl.JPAQueryFactory;
import jakarta.persistence.Tuple;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Repository;

@Repository
@RequiredArgsConstructor
public class EmailVerificationRepositoryImpl implements EmailVerificationRepository {
    private final EmailVerificationJpaRepository emailVerificationJpaRepository;
    private final JPAQueryFactory jpaQueryFactory;

    @Override
    public EmailVerification get(String token, EmailVerificationType emailVerificationType) {
        return emailVerificationJpaRepository.findByTokenAndEmailVerificationTypeAndIsVerifiedFalse(token, emailVerificationType)
            .orElseThrow(() -> new CustomException(Error.NOT_FOUND))
            .toModel();
    }

    @Override
    public EmailVerification getForNotVerifiedUser(Long userId) {
        EmailVerificationEntity result = jpaQueryFactory.select(emailVerificationEntity)
            .from(emailVerificationEntity)
            .where(
                emailVerificationEntity.uid.eq(userId),
                emailVerificationEntity.isVerified.isFalse(),
                emailVerificationEntity.emailVerificationType.eq(EmailVerificationType.CreateUser))
            .fetchFirst();

        if (result == null) {
            throw new CustomException(Error.NOT_FOUND);
        }

        return result.toModel();
    }

    @Override
    public EmailVerificationWithUser getForResend(String resendToken) {
        Tuple result = emailVerificationJpaRepository.findByResendTokenWithUser(resendToken)
            .orElseThrow(() -> new CustomException(Error.NOT_FOUND));
        EmailVerification emailVerification = result.get("emailVerification", EmailVerificationEntity.class).toModel();
        User user = result.get("user", UserEntity.class).toModel();
        return new EmailVerificationWithUser(emailVerification, user);
    }

    @Override
    public EmailVerification save(EmailVerification emailVerification) {
        return emailVerificationJpaRepository.save(emailVerification.toEntity()).toModel();
    }
}

File: qc-core/qc-user/qc-user-rdb/src/main/kotlin/com/eager/questioncloud/user/repository/EmailVerificationRepositoryImpl.kt
package com.eager.questioncloud.user.repository

import com.eager.questioncloud.common.exception.CoreException
import com.eager.questioncloud.common.exception.Error
import com.eager.questioncloud.user.domain.EmailVerification
import com.eager.questioncloud.user.entity.EmailVerificationEntity
import com.eager.questioncloud.user.entity.QEmailVerificationEntity.emailVerificationEntity
import com.eager.questioncloud.user.enums.EmailVerificationType
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.stereotype.Repository

@Repository
class EmailVerificationRepositoryImpl(
    private val emailVerificationJpaRepository: EmailVerificationJpaRepository,
    private val jpaQueryFactory: JPAQueryFactory,
) : EmailVerificationRepository {
    override fun get(token: String, emailVerificationType: EmailVerificationType): EmailVerification {
        return emailVerificationJpaRepository.findByTokenAndEmailVerificationTypeAndIsVerifiedFalse(
            token,
            emailVerificationType
        )
            .orElseThrow { CoreException(Error.NOT_FOUND) }
            .toModel()
    }
    
    override fun getCreateUserVerification(userId: Long): EmailVerification {
        return jpaQueryFactory.select(emailVerificationEntity)
            .from(emailVerificationEntity)
            .where(
                emailVerificationEntity.uid.eq(userId),
                emailVerificationEntity.isVerified.isFalse(),
                emailVerificationEntity.emailVerificationType.eq(EmailVerificationType.CreateUser)
            )
            .fetchFirst()?.toModel() ?: throw CoreException(Error.NOT_FOUND)
    }
    
    override fun getForResend(resendToken: String): EmailVerification {
        return jpaQueryFactory.select(emailVerificationEntity)
            .from(emailVerificationEntity)
            .where(
                emailVerificationEntity.isVerified.isFalse(),
                emailVerificationEntity.resendToken.eq(resendToken)
            )
            .fetchFirst()?.toModel() ?: throw CoreException(Error.NOT_FOUND)
    }
    
    override fun save(emailVerification: EmailVerification): EmailVerification {
        return emailVerificationJpaRepository.save(EmailVerificationEntity.createNewEntity(emailVerification)).toModel()
    }
    
    override fun update(emailVerification: EmailVerification) {
        emailVerificationJpaRepository.save(EmailVerificationEntity.fromExisting(emailVerification))
    }
}
